README.md
# ssh-tunnel
Establishes ssh tunnel for forwarding remote port to local one.

Original Author: Oleg Gaidarenko (markelog@gmail.com)


CHANGELOG.md
# Changelog

## 1.4.0 - 2017.06.27

* Display a kill signal with which tunnel is closed

## 1.3.0 - 2016.12.12

* Ability to create ssh tunnel via the specified port
* Add `user` public option

## 1.2.0 - 2016.09.30

* Add `host` & `port` public options

## 1.1.0 - 2016.08.31

* Add `user` option

## 1.0.0 - 2016.02.18

* Initial release


test/index.js
var Tunnel = require('../'),
    _ = require('lodash'),
    q = require('q'),
    childProcess = require('child_process'),
    events = require('events'),
    util = require('util');

describe('Tunnel', function () {
    var sandbox = sinon.sandbox.create(),
        tunnel,
        ssh;

    beforeEach(function () {
        ssh = new events.EventEmitter();
        ssh.stderr = new events.EventEmitter();
        ssh.kill = sandbox.stub();

        sandbox.stub(console, 'info');
        sandbox.stub(childProcess);
        childProcess.spawn.returns(ssh);
    });

    afterEach(function () {
        sandbox.restore();
    });

    describe('constructor', function () {
        it('should generate proxy host from host and ports range', function () {
            var tunnel = createTunnel({
                host: 'test_host',
                ports: { min: 0, max: 1 }
            });

            expect(tunnel.proxyHost).to.match(/test_host:\d+/);
        });

        it('should set port as passed in ports if min and max ports are same', function () {
            var tunnel = createTunnel({
                ports: { min: 8080, max: 8080 }
            });

            expect(tunnel.proxyHost).to.have.string(':8080');
        });

        it('should set default timeout as 10 seconds', function () {
            var tunnel = createTunnel(),
                timeout = sandbox.stub(q.makePromise.prototype, 'timeout'); // promise constructor available like this?!

            tunnel.open();

            expect(timeout).to.be.calledWith(10000);
        });
    });

    describe('instance methods', function () {
        describe('open', function () {
            it('should return promise', function () {
                tunnel = createTunnel();

                var result = tunnel.open();

                expect(q.isPromise(result)).to.be.true;
            });

            describe('tunnel spawn and events', function () {
                it('should spawn ssh process', function () {
                    tunnel = createTunnel();

                    tunnel.open();

                    expect(childProcess.spawn).to.be.calledWith('ssh');
                });

                it('should spawn tunnel from localhost to remote host for provided local and remote ' +
                    'ports', function () {
                    tunnel = createTunnel({
                        host: 'remote_host',
                        localport: 8080
                    });

                    tunnel.open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs[0]).to.match(/-R \d+:localhost:8080/);
                    expect(sshArgs).to.contain('remote_host')
                        .and.to.contain('-N')
                        .and.to.contain('-v');
                });

                it('should spawn tunnel via the default ssh port', function () {
                    createTunnel().open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('-p 22');
                });

                it('should spawn tunnel via the specified ssh port', function () {
                    createTunnel({ sshPort: 100500 }).open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('-p 100500');
                });

                it('should spawn tunnel to remote host using provided user ', function () {
                    tunnel = createTunnel({
                        host: 'remote_host',
                        user: 'user',
                        localport: 8080
                    });

                    tunnel.open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('user@remote_host');
                });

                it('should resolve promise if tunnel successfully created', function () {
                    tunnel = createTunnel();

                    var promise = tunnel.open();

                    ssh.stderr.emit('data', 'success');

                    return expect(promise).to.be.eventually.resolved;
                });

                it('should set resulting host and port to public variable', function () {
                    tunnel = createTunnel({
                        host: 'some_ssh_host'
                    });

                    tunnel.open();

                    ssh.stderr.emit('data', 'success');

                    expect(tunnel.proxyHost).to.match(/^some_ssh_host:\d+/);
                });

                it('should reject tunnel opening if failed to create tunnel', function () {
                    tunnel = createTunnel();

                    var openPromise = tunnel.open();

                    ssh.stderr.emit('data', 'failed');

                    expect(openPromise).to.be.rejected;
                });

                it('should do nothing if fail occured in stderr after successful tunnel opening', function () {
                    tunnel = createTunnel();

                    tunnel.open();
                    ssh.stderr.emit('data', 'success');
                    ssh.stderr.emit('data', 'failed');

                    expect(console.info)
                        .to.be.not.calledWith(util.format('ERROR: failed to create tunnel to %s.', tunnel.proxyHost));
                });

                it('should log a kill signal with which the tunnel is closed', function () {
                    tunnel = createTunnel();

                    tunnel.open();
                    ssh.stderr.emit('data', 'channel 0: \nKilled BY signal 2');

                    expect(console.info.secondCall.args[0])
                        .to.be.equal('INFO: Tunnel is killed by signal 2');
                });

                it('should not log a kill signal if tunnel is closed successfully', function () {
                    tunnel = createTunnel();

                    tunnel.open();
                    ssh.stderr.emit('data', 'Exit status 0');

                    expect(console.info).to.be.calledOnce;
                    expect(console.info.firstCall.args[0]).to.not.match(/INFO: Tunnel is killed/);
                });

                it('should reject tunnel opening if error occured', function () {
                    tunnel = createTunnel();

                    var openPromise = tunnel.open();

                    ssh.emit('error');

                    expect(openPromise).to.be.rejected;
                });

                it('should close tunnel if close event received', function () {
                    tunnel = createTunnel();
                    tunnel.open();

                    var closePromise = tunnel.close();

                    ssh.emit('close');

                    return expect(closePromise).to.be.eventually.resolved;
                });
            });
        });

        describe('close', function () {
            it('should return resolved promise if no tunnel opened', function () {
                tunnel = createTunnel();

                expect(tunnel.close()).to.eventually.resolved;
            });

            it('should return promise if tunnel opened', function () {
                tunnel = createTunnel();
                tunnel.open();

                expect(q.isPromise(tunnel.close())).to.be.true;
            });

            it('should try to kill tunnel using SIGTERM', function () {
                tunnel = createTunnel();
                tunnel.open();
                tunnel.close();

                expect(ssh.kill).to.be.calledWith('SIGTERM');
                expect(ssh.kill).to.be.not.calledWith('SIGKILL');
            });

            it('should try to kill tunnel with SIGKILL if tunnel was not closed in 3000ms after ' +
                'SIGTERM', function () {
                var clock = sinon.useFakeTimers(),
                    closePromise;

                tunnel = createTunnel();
                tunnel.open();

                closePromise = tunnel.close();
                expect(ssh.kill).to.be.calledWith('SIGTERM');

                clock.tick(3100);

                return closePromise.then(function () {
                    expect(ssh.kill).to.be.calledWith('SIGKILL');
                    clock.restore();
                });
            });
        });
    });

    describe('static methods', function () {
        describe('openWithRetries', function () {
            beforeEach(function () {
                sandbox.stub(Tunnel.prototype);
                Tunnel.prototype.open.returns(q());
                Tunnel.prototype.close.returns(q());
            });

            it('should return promise', function () {
                var result = Tunnel.openWithRetries(defaultOpts());

                expect(q.isPromise(result)).to.be.true;
            });

            it('should try to open tunnel with passed opts', function () {
                var opts = defaultOpts();

                Tunnel.openWithRetries(opts);

                expect(Tunnel.prototype.__constructor).to.be.calledWith(opts);
            });

            it('should resolve promise if tunnel opened successfully', function () {
                Tunnel.prototype.open.returns(q());

                return expect(Tunnel.openWithRetries(defaultOpts())).to.be.eventually.resolved;
            });

            it('should resolve promise with tunnel instance', function () {
                Tunnel.prototype.open.returns(q());

                return Tunnel.openWithRetries(defaultOpts()).then(function (tunnel) {
                    expect(tunnel).to.be.instanceOf(Tunnel);
                });
            });

            it('should reject tunnel if failed to open tunnel after retries', function () {
                Tunnel.prototype.open.returns(q.reject());

                return expect(Tunnel.openWithRetries(defaultOpts)).to.be.eventually.rejected;
            });

            it('should retry to create tunnel 5 times by default', function () {
                Tunnel.prototype.open.returns(q.reject());

                return Tunnel.openWithRetries(defaultOpts()).fail(function () {
                    expect(Tunnel.prototype.open.callCount).to.be.equal(5);
                });
            });

            it('should retry create tunnel retries times', function () {
                Tunnel.prototype.open.returns(q.reject());

                return Tunnel.openWithRetries(defaultOpts(), 10).fail(function () {
                    expect(Tunnel.prototype.open.callCount).to.be.equal(10);
                });
            });
        });
    });
});

function createTunnel (opts) {
    opts = opts || {};

    _.defaults(opts, defaultOpts());

    return new Tunnel(opts);
}

function defaultOpts () {
    return {
        host: 'deafault_host',
        ports: { min: 0, max: 1 },
        localport: 0
    };
}


test/setup.js
global.expect = require('chai').expect;
global.sinon = require('sinon');

require('chai')
    .use(require('sinon-chai'))
    .use(require('chai-as-promised'));


lib/tunnel.js
require('babel/polyfill');

import * as path from 'path';
import * as net from 'net';
import {readFileSync as read} from 'fs';

import Connection from 'ssh2';
import { extend } from 'lodash';
import * as vow from 'vow';

let defer = Symbol();

export default class Tunnel {

  /**
   * Define all needed properties
   * @param {Object} from
   * @param {String} from.hostname
   * @param {Number} from.port
   * @param {String} from.key - path to the user key
   * @param {Object} to
   * @param {String} to.hostname
   * @param {Number | Array} [to.port = 8000] - number or range
   * @param {String} to.username
   * @return {Tunnel}
   */
  constructor(from, to, logger) {

    /**
     * What hostname/port should we forward from?
     * @type {Object}
     */
    this.from = extend({}, from);

    /**
     * What hostname/port should we forward to?
     * @type {Object}
     */
    this.to = extend({}, to);

    /**
     * Original port range, needed if we infact received port range,
     * so we could extract new port for new connection
     * @type {Number | Array}
     */
    this.portRange = to.port;
    this.to.port = Tunnel.getPort(this.portRange);

    /**
     * Username of the remote host
     * @type {String}
     */
    this.username = to.username || process.env.USER;

    /**
     * Authorization key
     * @type {String}
     */
    this.key = read(from.key || path.join(process.env.HOME, '.ssh', 'id_rsa'));

    /**
     * Deferred object which we will resolve when connect to the remote host
     * @private
     * @type {Object}
     */
    this[defer] = vow.defer();

    /**
     * Promise object which will be resolved when connect to the remote host
     * @type {Object}
     */
    this.promise = this[defer].promise();

    /**
     * Logger that we will use for output info
     * @type {Object}
     */
    this.logger = Tunnel.getLogger(logger);

    /**
     * Our ssh connection
     * @type {Object}
     */
    this.connection = new Connection();

    /**
     * How many times should we try to reconnect?
     * @type {Number}
     */
    this.retryTimes = 0;
  }

  /**
   * Connect to the remote server
   * @param {Number} [retryTimes = 0] - how many times can we try to connect
   */
  connect(retryTimes = 0) {
    this.retryTimes = retryTimes;

    this.connection.connect({
      host: this.to.hostname,
      port: 22,
      username: this.username,
      privateKey: this.key
    });
  }

  /**
   * Close connection to the remote server
   */
  close() {
    this.connection.end();
  }

   /**
   * Helper for the Tunnel#reTry
   * @private
   */
  _reTry() {
    // Close current connection
    this.close();

    // Try new port
    this.to.port = Tunnel.getPort(this.portRange);

    // Recreate the connection
    this.connection = new Connection();

    this.addEvents();

    // Decrement the shot
    this.connect(this.retryTimes - 1);
  }

  /**
   * Try to reconnect to the remote host (has side effect)
   * @param {*} error - error that will be thrown if we don't want to try anymore
   */
  reTry(error) {
    if (this.retryTimes !== 0) {
      this.logger.info('Retrying to connect, %s tries left', this.retryTimes);
      this._reTry();

    } else {
      this[defer].reject(error);
    }
  }

  /**
   * Attach events and deal with them
   */
  addEvents() {
    var streams = [];

    this.connection.on('tcp connection', (info, accept, reject) => {
      let socket;
      let stream = accept();

      stream.pause();

      // Connect to the socket and output the stream
      socket = net.connect(this.from.port, this.from.hostname, () => {
        stream.pipe(socket);
        socket.pipe(stream);

        stream.resume();
      });

      // Store all streams, so we can clean it up afterwards
      streams.push(socket, stream);
    });

    // Wait for remote connection
    this.connection.on('ready', () => {
      this.logger.info('Connection to %s:%s is established', this.to.hostname, this.to.port);

      // Forward all connections to remote address from local hostname and port
      this.connection.forwardIn(this.to.hostname, this.to.port, (error) => {
        if (error) {
          this.logger.error('Forwarding issue %s', error.message);

        } else {
          this[defer].resolve();
        }
      });
    });

    this.connection.on('error', (error) => {
      this.logger.error('Connection error %s', error.message);
      this.reTry();
    });

    this.connection.on('close', (error) => {
      // End all streams, to clean up event loop queue
      streams.forEach((stream) => stream.end());

      if (error) {
        this.logger.error('Connection error %s', error.message);

      } else {
        this.logger.info('Connection closed');
      }
    });
  }

  /**
   * Return logger or return default one is there is none
   * @static
   * @param {Object} [logger]
   * @return {Object}
   */
  static getLogger(logger) {
    if (logger) {
      return logger;
    }

    return {
      info: console.log.bind(console),
      error: console.error.bind(console)
    };
  }

  /**
   * Return default port or passed port or get port within the range
   * @static
   * @param {Number | Array} [port = 8000]
   * @return {Number}
   */
  static getPort(port) {
    port = port || 8000;

    if (Array.isArray(port)) {
      port = Tunnel.getRandom(port[0], port[1]);
    }

    return port;
  }

  /**
   * Get random number
   * @static
   * @param {Number} min
   * @param {Number} max
   * @return {Number}
   */
  static getRandom(min, max) {
    return Math.floor(Math.random() * (max - min)) + min;
  }
}


index.js
'use strict';

var inherit = require('inherit'),
    q = require('q'),
    childProcess = require('child_process'),
    util = require('util');

var DEFAULTS = {
    MAX_RETRIES: 5,
    CONNECT_TIMEOUT: 10000,
    SSH_PORT: 22
};

var Tunnel = inherit({
    /**
     * Constuctor
     * @param {object} opts tunnel options
     * @param {string} opts.host remote host address
     * @param {object} opts.ports remote host ports range
     * @param {number} opts.localport local port number
     * @param {string} [opts.user] remote host user
     * @param {string} [opts.sshPort=22] host port to create tunnel
     * @param {number} [opts.maxRetries=5] max attempts to create tunnel
     * @param {number} [opts.connectTimeout=10000] ssh connect timeout
     */
    __constructor: function (opts) {
        this.host = opts.host;
        this.port = this._generateRandomPort(opts.ports);
        this._sshPort = opts.sshPort || DEFAULTS.SSH_PORT;
        this.user = opts.user;
        this.proxyHost = util.format('%s:%d', this.host, this.port);
        this.proxyUrl = this.proxyHost; // deprecated, use proxyHost
        this._localPort = opts.localport;
        this._connectTimeout = opts.connectTimeout || DEFAULTS.CONNECT_TIMEOUT;
        this._tunnel = null;
        this._tunnelDeferred = q.defer();
        this._closeDeferred = q.defer();
    },

    /**
     * Tries to open ssh connection to remote server
     * @returns {Promise}
     */
    open: function () {
        var _this = this;

        console.info('INFO: creating tunnel to %s', this.proxyHost);

        this._tunnel = childProcess.spawn('ssh', this._buildSSHArgs());

        this._tunnel.stderr.on('data', function (data) {
            if (/success/.test(data)) {
                return _this._resolveTunnel();
            }

            if (/failed/.test(data)) {
                if (_this._tunnelDeferred.promise.isFulfilled()) {
                    return;
                }
                return _this._rejectTunnel();
            }

            if (/killed/i.test(data)) {
                var msg = data.toString().toLowerCase();
                var killMsg = msg.slice(msg.indexOf('killed')).trim();

                console.info('INFO: Tunnel is ' + killMsg);
            }
        });

        this._tunnel.on('close', function (code) {
            return _this._closeTunnel(code);
        });

        this._tunnel.on('error', function () {
            return _this._rejectTunnel();
        });

        return _this._tunnelDeferred.promise.timeout(this._connectTimeout);
    },

    /**
     * Closes connection. If no connection established does nothing
     * @returns {Promise}
     */
    close: function () {
        if (!this._tunnel) {
            return q();
        }

        var _this = this;

        this._tunnel.kill('SIGTERM');
        return this._closeDeferred.promise.timeout(3000).fail(function () {
            _this._tunnel.kill('SIGKILL');
            return _this._closeTunnel(-1);
        });
    },

    _resolveTunnel: function () {
        console.info('INFO: Tunnel created to %s', this.proxyHost);
        this._tunnelDeferred.resolve();
    },

    _rejectTunnel: function () {
        var message = util.format('ERROR: failed to create tunnel to %s.', this.proxyHost),
            error = new Error(message);

        console.info(message);
        this._tunnelDeferred.reject(error);
    },

    _closeTunnel: function (exitCode) {
        console.info('INFO: Tunnel to %s closed. Exit code: %d', this.proxyHost, exitCode);
        this._closeDeferred.resolve();
    },

    _buildSSHArgs: function () {
        return [
            util.format('-R %d:localhost:%d', this.port, this._localPort),
            '-N',
            '-v',
            util.format('-p %d', this._sshPort),
            (this.user ? this.user + '@' : '') + this.host
        ];
    },

    _generateRandomPort: function (ports) {
        var min = ports.min,
            max = ports.max;

        return Math.floor(Math.random() * (max - min + 1)) + min;
    }
});

/**
 * Tries to open tunnel several times
 * @param {object} opts opts which will be passed to created tunnel
 * @param {number} retries amount of retries to open tunnel
 * @returns {Promise}
 */
Tunnel.openWithRetries = function (opts, retries) {
    retries = retries || DEFAULTS.MAX_RETRIES;

    function retry_(retriesLeft) {
        if (!retriesLeft) {
            return q.reject(util.format('ERROR: failed to create tunnel after %d attempts', retries));
        }

        var tunnel = new Tunnel(opts);

        return tunnel.open()
            .then(function () {
                return q.resolve(tunnel);
            })
            .fail(function () {
                return tunnel.close()
                    .then(retry_.bind(null, retriesLeft - 1));
            });
    }

    return retry_(retries);
};

module.exports = Tunnel;
